File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.cpp
class Solution {
public:
    int MOD = 1000000007; 

    int solve(int d, int f, int target, vector<vector<int>>& memo) {
      
        if (d == 0 && target == 0) return 1;
        if (d == 0 || target < 0) return 0;
        if (memo[d][target] != -1) return memo[d][target];

        int sum = 0;
        for (int i = 1; i <= f; i++) {
            sum = (sum % MOD + solve(d - 1, f, target - i, memo) % MOD) % MOD;
        }

        memo[d][target] = sum;
        return sum;
    }

    int numRollsToTarget(int d, int f, int target) {
        vector<vector<int>> memo(d + 1, vector<int>(target + 1, -1));
        return solve(d, f, target, memo);
    }
};

File: 1155-number-of-dice-rolls-with-target-sum/1155-number-of-dice-rolls-with-target-sum.java
class Solution {
    private static final int MOD = (int)1e9 + 7;

    private int solve(int n, int k, int target, int[][] dp) {
        if (target < 0) {
            return 0;
        }

        if (n == 0) {
            return target == 0 ? 1 : 0;
        }

        if (dp[n][target] != -1) {
            return dp[n][target];
        }

        int ways = 0;
        for (int i = 1; i <= k; i++) {
            ways = (int)((ways + solve(n - 1, k, target - i, dp)) % MOD);
        }
        dp[n][target] = ways;
        return dp[n][target];
    }

    public int numRollsToTarget(int n, int k, int target) {
        int[][] dp = new int[n + 1][target + 1];
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }
        return solve(n, k, target, dp);
    }
}
